src/linker/Linker/SignatureParser.cs

```diff
@@ -0,0 +1,661 @@
+
+					// has type parameters?
+					if (PeekNextChar (id, index) == '`') {
+						index++;
+
+						bool genericType = true;
+
+						// method type parameters?
+						// note: this allows `` for type parameters
+						if (PeekNextChar (id, index) == '`') {
+							index++;
+							genericType = false;
+						}
+
+						arity = ReadNextInteger (id, ref index);
+
+						if (genericType) {
+							// We need to mangle generic type names but not generic method names.
+							nameBuilder.Append ('`');
+							nameBuilder.Append (arity);
+						}
+					}
+
+					// no more dots, so don't loop any more
+					if (PeekNextChar (id, index) != '.')
+						break;
+
+					// must be a namespace or type since name continues after dot
+					index++;
+
+					// try to resolve it as a type
+					var typeOrNamespaceName = nameBuilder.ToString ();
+					GetMatchingTypes (module, declaringType: containingType, name: typeOrNamespaceName, results: results);
+					Debug.Assert (results.Count <= 1);
+					if (results.Any ()) {
+						// the name resolved to a type
+						var result = results.Single ();
+						Debug.Assert (result is TypeDefinition);
+						// result becomes the new container
+						containingType = result as TypeDefinition;
+						nameBuilder.Clear ();
+						results.Clear ();
+						continue;
+					}
+
+					// it didn't resolve as a type.
+
+					// only types have arity.
+					if (arity > 0)
+						return;
+
+					// treat it as a namespace and continue building up the type name
+					nameBuilder.Append ('.');
+				}
+
+				if (containingType == null && memberType != MemberType.Type)
+					return;
+
+				var memberName = nameBuilder.ToString ();
+
+				switch (memberType) {
+				case MemberType.Method:
+					GetMatchingMethods (id, ref index, containingType, memberName, arity, results);
+					break;
+				case MemberType.Type:
+					GetMatchingTypes (module, containingType, memberName, results);
+					break;
+				case MemberType.Property:
+					GetMatchingProperties (id, ref index, containingType, memberName, results);
+					break;
+				case MemberType.Event:
+					GetMatchingEvents (containingType, memberName, results);
+					break;
+				case MemberType.Field:
+					GetMatchingFields (containingType, memberName, results);
+					break;
+				}
+			}
+		}
+
+		// Roslyn resolves types in a signature to their declaration by searching through namespaces.
+		// To avoid looking for types by name in all referenced assemblies, we just represent types
+		// that are part of a signature by their doc comment strings, and we check for matching
+		// strings when looking for matching member signatures.
+		private static string? ParseTypeSymbol (string id, ref int index, IGenericParameterProvider? typeParameterContext)
+		{
+			var results = new List<string> ();
+			ParseTypeSymbol (id, ref index, typeParameterContext, results);
+			if (results.Count == 0) {
+				return null;
+			} else {
+				Debug.Assert (results.Count == 1);
+				return results[0];
+			}
```

Comment:
I'd invert the logic if you are expecting `results.Count == 1` to be the common path